python/perf-kernels/03-matrix-multiplication-stream-k.py

```diff
@@ -0,0 +1,407 @@
+from typing import Optional
+
+import torch
+import triton
+import triton.language as tl
+import random
+
+#from triton.runtime.driver import CudaUtils
+import json
+
+torch.manual_seed(123)
+random.seed(123)
+
+#device = torch.cuda.current_device()
+#cuda_utils = CudaUtils()
+#total_sm = cuda_utils.get_device_properties(device)["multiprocessor_count"]
+total_sm = 110 # for MI250
+print(f"total SMs: {total_sm}")
+
+# ---------------------------------------------------------------------------
+# Triton kernels
+# ---------------------------------------------------------------------------
+
+
+@triton.jit()
+def swizzle_tile(tile_id,
+                 M, N, K,
+                 BLOCK_M: tl.constexpr, BLOCK_N: tl.constexpr, BLOCK_K: tl.constexpr,
+                 GROUP_M: tl.constexpr
+                 ):
+    grid_m = tl.cdiv(M, BLOCK_M)
+    grid_n = tl.cdiv(N, BLOCK_N)
+    # re-order program ID for better L2 performance
+    width = GROUP_M * grid_n
+    group_id = tile_id // width
+    group_size = tl.minimum(grid_m - group_id * GROUP_M, GROUP_M)
+    pid_m = group_id * GROUP_M + (tile_id % group_size)
+    pid_n = (tile_id % width) // group_size
+    return pid_m, pid_n
+
+
+@triton.jit()
+def linear_tile(tile_id,
+                M, N, K,
+                BLOCK_M: tl.constexpr, BLOCK_N: tl.constexpr, BLOCK_K: tl.constexpr,
+                GROUP_M: tl.constexpr
+                ):
+    pid_m = tile_id // tl.cdiv(N, BLOCK_N)
+    pid_n = tile_id % tl.cdiv(N, BLOCK_N)
+    return pid_m, pid_n
+
+
+# iterate, multiply and accumulate over K axis
+@triton.jit()
+def mac_loop(A, B, C,
+             M, N, K,
+             locks,
+             stride_am, stride_ak, stride_bk, stride_bn, stride_cm, stride_cn,
+             iters_per_tile,
+             start_iter, end_iter,
+             BLOCK_M: tl.constexpr, BLOCK_N: tl.constexpr, BLOCK_K: tl.constexpr,
+             ACC_TYPE: tl.constexpr, GROUP_M: tl.constexpr):
+
+    # where are we in the grid
+    tile_id = start_iter // iters_per_tile
+    if GROUP_M  > 0:
+        pid_m, pid_n = swizzle_tile(tile_id, M, N, K, BLOCK_M, BLOCK_N, BLOCK_K, GROUP_M)
+    else:
+        pid_m, pid_n = linear_tile(tile_id, M, N, K, BLOCK_M, BLOCK_N, BLOCK_K, GROUP_M)
+
+    rm = pid_m * BLOCK_M + tl.arange(0, BLOCK_M)
+    rn = pid_n * BLOCK_N + tl.arange(0, BLOCK_N)
+    rk = tl.arange(0, BLOCK_K)
+    A = A + rm[:, None] * stride_am + rk[None, :] * stride_ak + BLOCK_K * stride_ak * (start_iter % iters_per_tile)
+    B = B + rk[:, None] * stride_bk + rn[None, :] * stride_bn + BLOCK_K * stride_bk * (start_iter % iters_per_tile)
+    acc = tl.zeros((BLOCK_M, BLOCK_N), dtype=ACC_TYPE)
+
+    for current_iter in range(start_iter, end_iter):
+        a = tl.load(A)
+        b = tl.load(B)
+        acc += tl.dot(a, b)
+        A += BLOCK_K * stride_ak
+        B += BLOCK_K * stride_bk
+
+    if end_iter % iters_per_tile == 0:  # last iteration of the tile always happens before its start on another SM
```

Comment:
Is there a way to avoid this modulus? The way we do it in our version is;
```cpp
if (iter != tileIter) {
...
else {
...
}
```

Where `iter` is `current_iter` and `tileIter` is `tileId * itersPerTile`, this should be cheaper than mod.

Reply:
with modulus:
(py_3.9) root@banff-cyxtera-s71-2:/home/work/triton/python/perf-kernels# python 03-matrix-multiplication-stream-k.py
total SMs: 304
M,N,K=8192,8192,8192 ; BLK_M,N,K=128,128,32
total_blocks_M=64 x total_blocks_N=64 = total_tiles=4096
total_tiles_streamk=448 + total_blocking_tiles=3648 = total_tiles=4096
total_programs_streamk=304
total_blocking_tiles=3648
iters_per_tile=256
total_iters_streamk=114688
0 registers used, 0 spills
0 registers used, 0 spills
PyTorch: 3.062 ms  359.134 tflops
hybrid stream-k (grid=304): 5.661 ms  194.233 tflops
hybrid stream-k (grid=608): 5.852 ms  187.887 tflops
tile matmul (grid=0): 5.590 ms  196.683 tflops

without modules:
(py_3.9) root@banff-cyxtera-s71-2:/home/work/triton/python/perf-kernels# python 03-matrix-multiplication-stream-kopt.py
total SMs: 304
M,N,K=8192,8192,8192 ; BLK_M,N,K=128,128,32
total_blocks_M=64 x total_blocks_N=64 = total_tiles=4096
total_tiles_streamk=448 + total_blocking_tiles=3648 = total_tiles=4096
total_programs_streamk=304
total_blocking_tiles=3648
iters_per_tile=256
total_iters_streamk=114688
0 registers used, 0 spills
0 registers used, 0 spills
PyTorch: 3.051 ms  360.402 tflops
hybrid stream-k (grid=304): 5.665 ms  194.093 tflops
hybrid stream-k (grid=608): 5.848 ms  188.005 tflops
tile matmul (grid=0): 5.588 ms  196.754 tflops

see b614dcb41..4de0c9df5 for code difference

Reply:
I just realized I haven't reduced the number of usage of modulus, I will change it soon

Reply:
Yeah, it seems like they were moved elsewhere. You can completely avoid them instead. :)